src/output/gtk-shell.cpp

```diff
@@ -2,77 +2,205 @@
 #include "gtk-shell-protocol.h"
 
 #include <wayfire/util/log.hpp>
-#include "wayfire/view.hpp"
+#include <wayfire/view.hpp>
 #include "../core/core-impl.hpp"
-#include "wayfire/core.hpp"
+#include <wayfire/core.hpp>
 #include <map>
+#include <wayfire/signal-definitions.hpp>
+#define GTK_SHELL_VERSION 3
 
-struct wf_gtk_shell {
+struct wf_gtk_shell
+{
     std::map<wl_resource*, std::string> surface_app_id;
 };
 
-static void handle_gtk_surface_set_dbus_properties(wl_client *, wl_resource *resource,
-    const char *application_id,
-    const char *, const char *, const char *, const char *, const char *)
+
+/**
+ *  In gnome-shell/mutter/meta windows/views keep track of the properties
+ *  specified as arguments here.
+ *  Currently only the app_id is implemented / required.
+ */
+static void handle_gtk_surface_set_dbus_properties(wl_client *client, wl_resource *resource,
+    const char *application_id, const char *app_menu_path, const char *menubar_path,
+    const char *window_object_path, const char *application_object_path,
+    const char *unique_bus_name)
 {
     auto surface = static_cast<wl_resource*> (wl_resource_get_user_data(resource));
     if (application_id)
         wf::get_core_impl().gtk_shell->surface_app_id[surface] = application_id;
 }
 
-static void handle_gtk_surface_set_modal(wl_client *client, wl_resource *resource) { }
-static void handle_gtk_surface_unset_modal(wl_client *client, wl_resource *resource) { }
-static void handle_gtk_surface_present(wl_client *client, wl_resource *resource, uint32_t time) { }
-static void handle_gtk_surface_destroy(wl_resource *resource) { }
 
+/**
+ * Modal dialogs may be handled differently than non-modal dialogs.
+ * It is a hint that this should be attached to the parent surface.
+ * In gnome this does not affect input-focus.
+ * This function sets the modal hint.
+ */
+static void handle_gtk_surface_set_modal(wl_client *client, wl_resource *resource)
+{
+    auto surface = static_cast<wl_resource*> (wl_resource_get_user_data(resource));
+    wayfire_view view = wf::wl_surface_to_wayfire_view(surface);
+    if (view && view != nullptr)
```

Comment:
```suggestion
    if (view)
```
Seems like you're doing the same check twice?